File: ios/Sources/MritMesh/Core/MeshID.swift
import Foundation
import CryptoKit

/// MeshID — unique 256-bit identifier for every node on the MRIT mesh network.
///
/// Generated once on first install. Not a phone number. Not an IP. Ours.
///
/// Generation:  SHA-256(EC_P256_PublicKey + InstallTimestamp + RandomSalt)
/// Wire format: 32 raw bytes — identical to Android's MeshID.
public struct MeshID: Hashable, Sendable {

    /// Raw 32-byte (256-bit) identifier
    public let bytes: Data

    public init(bytes: Data) {
        precondition(bytes.count == 32, "MeshID must be exactly 32 bytes")
        self.bytes = bytes
    }

    // ── Identity ───────────────────────────────────────────────────────────────

    /// Full 64-character lowercase hex string
    public var hexString: String {
        bytes.map { String(format: "%02x", $0) }.joined()
    }

    /// Short 8-character uppercase hex — used in UI and logs
    public var shortId: String {
        String(hexString.prefix(8)).uppercased()
    }

    // ── Factory methods ────────────────────────────────────────────────────────

    /// Generate a new MeshID for this device.
    /// Called exactly once on first launch — result stored in Keychain.
    public static func generate() -> MeshID {
        // EC P-256 key pair — same curve as Android
        let privateKey  = P256.KeyAgreement.PrivateKey()
        let pubKeyBytes = privateKey.publicKey.x963Representation  // 65-byte uncompressed point

        let timestampBytes = withUnsafeBytes(of: Int64(Date().timeIntervalSince1970 * 1000).bigEndian) { Data($0) }

        var salt = Data(count: 32)
        salt.withUnsafeMutableBytes { _ = SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }

        var hasher = SHA256()
        hasher.update(data: pubKeyBytes)
        hasher.update(data: timestampBytes)
        hasher.update(data: salt)
        return MeshID(bytes: Data(hasher.finalize()))
    }

    /// Reconstruct a MeshID from a 64-character hex string.
    public static func fromHex(_ hex: String) -> MeshID? {
        guard hex.count == 64 else { return nil }
        var result = Data(count: 32)
        for i in 0..<32 {
            let lo = hex.index(hex.startIndex, offsetBy: i * 2)
            let hi = hex.index(lo, offsetBy: 2)
            guard let byte = UInt8(hex[lo..<hi], radix: 16) else { return nil }
            result[i] = byte
        }
        return MeshID(bytes: result)
    }

    /// Reconstruct a MeshID from raw bytes (e.g. from a received packet header).
    public static func fromBytes(_ data: Data) -> MeshID? {
        guard data.count >= 32 else { return nil }
        return MeshID(bytes: data.prefix(32))
    }

    /// Broadcast address — packet addressed to ALL nearby nodes (all 0xFF bytes).
    public static let broadcast = MeshID(bytes: Data(repeating: 0xFF, count: 32))
}

extension MeshID: CustomStringConvertible {
    public var description: String { hexString }
}

File: app/src/main/java/com/mrit/mesh/core/MeshID.kt
package com.mrit.mesh.core

import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * MeshID — unique 256-bit identifier for every node on the MRIT mesh network.
 *
 * Generated once on first install. Not a phone number. Not an IP. Ours.
 *
 * Generation formula:
 *   MeshID = SHA-256(EC_PublicKey_bytes + InstallTimestamp + RandomSalt)
 *
 * This ensures:
 *   - Globally unique (EC key pair + salt + time)
 *   - Not traceable to personal identity (no phone number, no IMEI)
 *   - 32 bytes / 256-bit — compact for packet headers
 */
data class MeshID(val bytes: ByteArray) {

    init {
        require(bytes.size == 32) { "MeshID must be exactly 32 bytes (256-bit)" }
    }

    /** Full hex string, e.g. "a3f79c12...8b41" (64 chars) */
    override fun toString(): String = bytes.joinToString("") { "%02x".format(it) }

    /** Short display name for UI — first 8 hex chars */
    fun shortId(): String = toString().take(8).uppercase()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MeshID) return false
        return bytes.contentEquals(other.bytes)
    }

    override fun hashCode(): Int = bytes.contentHashCode()

    companion object {

        /**
         * Generate a new MeshID for this device.
         * Called exactly once on first app install — result stored in encrypted SharedPreferences.
         */
        fun generate(): MeshID {
            // Generate EC P-256 key pair (Curve25519 requires BouncyCastle on older Android)
            val keyGen = KeyPairGenerator.getInstance("EC")
            keyGen.initialize(256, SecureRandom())
            val publicKeyBytes = keyGen.generateKeyPair().public.encoded

            val timestamp = System.currentTimeMillis().toString().toByteArray(Charsets.UTF_8)
            val salt = ByteArray(32).also { SecureRandom().nextBytes(it) }

            val digest = MessageDigest.getInstance("SHA-256")
            digest.update(publicKeyBytes)
            digest.update(timestamp)
            digest.update(salt)

            return MeshID(digest.digest())
        }

        /**
         * Restore a MeshID from its stored hex string.
         *
         * @param hex 64-character hex string (e.g. from SharedPreferences)
         */
        fun fromHex(hex: String): MeshID {
            require(hex.length == 64) { "Hex must be 64 characters, got ${hex.length}" }
            val bytes = ByteArray(32) { i ->
                hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
            }
            return MeshID(bytes)
        }

        /** Reconstruct MeshID from raw 32-byte array (e.g. from packet header). */
        fun fromBytes(bytes: ByteArray): MeshID {
            require(bytes.size >= 32) { "Need at least 32 bytes, got ${bytes.size}" }
            return MeshID(bytes.copyOf(32))
        }

        /**
         * Broadcast address — packet addressed to ALL nearby nodes.
         * All 0xFF bytes: FF FF FF ... FF (32 bytes)
         */
        val BROADCAST: MeshID = MeshID(ByteArray(32) { 0xFF.toByte() })
    }
}
